File: 2914-find-the-safest-path-in-a-grid/find-the-safest-path-in-a-grid.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:

    bool canReach(vector<vector<int>>& dist, int safe) {
        int n = dist.size();

        if (dist[0][0] < safe) return false;

        queue<pair<int,int>> q;
        vector<vector<bool>> vis(n, vector<bool>(n,false));

        q.push({0,0});
        vis[0][0] = true;

        int dir[4][2] = {{1,0},{-1,0},{0,1},{0,-1}};

        while (!q.empty()) {
            auto [r,c] = q.front();
            q.pop();

            if (r == n-1 && c == n-1)
                return true;

            for (auto &d : dir) {
                int nr = r + d[0];
                int nc = c + d[1];

                if (nr>=0 && nc>=0 && nr<n && nc<n &&
                    !vis[nr][nc] &&
                    dist[nr][nc] >= safe) {

                    vis[nr][nc] = true;
                    q.push({nr,nc});
                }
            }
        }
        return false;
    }

    int maximumSafenessFactor(vector<vector<int>>& grid) {

        int n = grid.size();

        vector<vector<int>> dist(n, vector<int>(n, INT_MAX));
        queue<pair<int,int>> q;

        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j]==1){
                    dist[i][j]=0;
                    q.push({i,j});
                }
            }
        }

        int dir[4][2]={{1,0},{-1,0},{0,1},{0,-1}};

        while(!q.empty()){
            auto [r,c]=q.front();
            q.pop();

            for(auto &d:dir){
                int nr=r+d[0];
                int nc=c+d[1];

                if(nr>=0 && nc>=0 && nr<n && nc<n &&
                   dist[nr][nc]==INT_MAX){

                    dist[nr][nc]=dist[r][c]+1;
                    q.push({nr,nc});
                }
            }
        }

        int low=0, high=2*n, ans=0;

        while(low<=high){
            int mid=(low+high)/2;

            if(canReach(dist,mid)){
                ans=mid;
                low=mid+1;
            }
            else{
                high=mid-1;
            }
        }

        return ans;
    }
};

File: 2914-find-the-safest-path-in-a-grid/find-the-safest-path-in-a-grid.py
from collections import deque

class Solution(object):

    def canReach(self, dist, safe):
        n = len(dist)

        if dist[0][0] < safe:
            return False

        q = deque([(0, 0)])
        visited = set()
        visited.add((0, 0))

        directions = [(1,0), (-1,0), (0,1), (0,-1)]

        while q:
            r, c = q.popleft()

            if r == n-1 and c == n-1:
                return True

            for dr, dc in directions:
                nr, nc = r+dr, c+dc

                if (0 <= nr < n and
                    0 <= nc < n and
                    (nr, nc) not in visited and
                    dist[nr][nc] >= safe):

                    visited.add((nr, nc))
                    q.append((nr, nc))

        return False


    def maximumSafenessFactor(self, grid):
        n = len(grid)

        dist = [[float('inf')] * n for _ in range(n)]
        q = deque()

        for i in range(n):
            for j in range(n):
                if grid[i][j] == 1:
                    dist[i][j] = 0
                    q.append((i, j))

        directions = [(1,0), (-1,0), (0,1), (0,-1)]

        while q:
            r, c = q.popleft()

            for dr, dc in directions:
                nr, nc = r+dr, c+dc

                if (0 <= nr < n and
                    0 <= nc < n and
                    dist[nr][nc] == float('inf')):

                    dist[nr][nc] = dist[r][c] + 1
                    q.append((nr, nc))

        low, high = 0, 2*n
        ans = 0

        while low <= high:
            mid = (low + high) // 2

            if self.canReach(dist, mid):
                ans = mid
                low = mid + 1
            else:
                high = mid - 1

        return ans